swarm/src/lib.rs

```diff
@@ -802,6 +891,7 @@ where
                             other_established: other_established_connection_ids.len(),
                         },
                     ));
+                self.supported_protocols = supported_protocols;
```

Comment:
I don't think we can do this in a separate PR. It definitely wasn't possible in earlier iterations. Will check.

The problem is that `new_handler` will panic by default now because you are not meant to call it. Thus, I cannot rely on it anywhere but have to call the new callbacks.

Those require connection-specific metadata that I only have after one is established.

Reply:
> Could this cause any side-effects for folks depending on the previous behavior? Would you mind introducing this change in a separate pull request?

Previously, we gathered the supported protocols on start-up of the swarm. Now, we are gathering it on the first connection. I think this can be observable if a `NetworkBehaviour` expects to read all supported protocols on the first call to `NetworkBehaviour::poll` via `PollParameters` when no connection is established yet.

I don't see a way around this change though unless we want to first completely redesign how `NetworkBehaviour`s signal which protocols they support.

We already agreed that `PollParameters` needs to go away (https://github.com/libp2p/rust-libp2p/issues/3124). I think the answer to this is that as part of https://github.com/libp2p/rust-libp2p/issues/2680, we scope supported protocols to a connection and thus remove this Swarm-wide state of "supported protocols".

Reply:
Thanks for elaborating. Let's keep as is :+1: